### PR TITLE
Tool to tune system parameters like PoH service priority

### LINKED ISSUE #6482: PoH service takes variable amount of time to generate hashes
#### Problem
PoH service generates a certain number of hashes (configured via Genesis block, as hashes_per_tick) before registering a tick. The service thread sets the CPU affinity, but due to OS scheduling, there is no guarantee that no other thread will get scheduled on the CPU core. This causes different node types (leader/validator) to take different amount of time to compute the same number of hashes. This causes network drift as nodes complete their respective slots at different times. End result can be leader timeouts, and dropped slots.

#### Proposed Solution
I tried setting the scheduling policy for PoH service thread to realtime, with FIFO policy. This helps align the PoH time for all nodes. 

Running a 5 node network, I plotted the time it took to do 8 ticks on the cluster.
The following graphs is for tip of master
image

This graph is with the scheduling change
image

It can be seen that PoH ticks are more consistent with the change.

#### Challenges
1. The thread itself is not able to set it's scheduling policy due to lack of privileges. It needs superuser privileges. I tried using `thread_priority` crate, and ran into the permissions issue.
2. We can use `chrt` to update the scheduling policy from shell. It has to be run using sudo. The above graphs were captured using this approach. I added the following code to remote-node.sh
```sudo chrt -r -p 99 `ps -eT | grep solana-poh-serv | awk '{print $2}'```
This approach can not be used for external nodes, as the node boots up directly from the native `solana-validator` program.
3. Just changing the priority of the thread (using `renice`) did not help with the problem.

### PR DESCRIPTION
**Pull request by pgarg66**
#### Problem
The PoH service thread scheduling policy and priority cannot be tuned unless we use a shell script. Using the shell script is not a viable option for external nodes.

#### Summary of Changes
Added a new daemon `sys_tuner` that runs at a higher privilege. It listens on a local domain socket. When validator boots up, the `poh_service` connects to the socket. It triggers the `sys_tuner` to look up `poh_service's` thread ID, and adjust its priority and scheduling policy.

Fixes #6482

### CODE REVIEW

File: net/remote/remote-node.sh [hunk 1]
```diff
@@ -129,6 +129,8 @@ cat >> ~/solana/on-reboot <<EOF
   # shellcheck source=/dev/null
   SUDO_OK=1 source scripts/tune-system.sh
 
+  sudo solana-sys-tuner &
+
```
Comment (mvines) [inline net/remote/remote-node.sh]: Please capture the pid into a .pid file so `solana-sys-tuner` gets killed on a `./net.sh stop`

File: sys-tuner/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,8 @@
+use crate::main::SOLANA_SYS_TUNER_PATH;
+use unix_socket::UnixStream;
+
+pub mod main;
+
+pub fn request_system_tuning() {
```
Comment (mvines) [inline sys-tuner/src/lib.rs]: wdyt about the name `request_realtime_poh()`

File: sys-tuner/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+license = "Apache-2.0"
+homepage = "https://solana.com/"
+publish = true
+
+[dependencies]
+clap = "2.33.0"
+log = "0.4.8"
+semver = "0.9.0"
+solana-logger = { path = "../logger", version = "0.22.0" }
+unix_socket2 = "0.5.4"
```
Comment (mvines) [inline sys-tuner/Cargo.toml]: `unix_socker2` should be under `[target."cfg(unix)".dependencies]`, windows doesn't have it.

eg, https://github.com/solana-labs/solana/blob/5a30ef180abf0c87a01f2b17b5add4b5af7fe6ad/core/Cargo.toml#L74

File: sys-tuner/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,62 @@
+use std::fs;
+use unix_socket::UnixListener;
```
Comment (mvines) [inline sys-tuner/src/main.rs]: Wrap in `#[cfg(unix)]` (for windows)

File: sys-tuner/src/main.rs [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+            }
+        }
+    }
+}
+
+#[allow(dead_code)]
+#[cfg(any(not(unix), target_os = "macos"))]
+fn tune_system() {}
+
+#[allow(dead_code)]
```
Comment (mvines) [inline sys-tuner/src/main.rs]: What's all the `#[allow(dead_code)]` about?

File: sys-tuner/src/main.rs [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+fn main() {
+    solana_logger::setup();
+    let listener = match UnixListener::bind(SOLANA_SYS_TUNER_PATH) {
+        Ok(l) => l,
+        Err(_) => {
+            fs::remove_file(SOLANA_SYS_TUNER_PATH).unwrap();
+            UnixListener::bind(SOLANA_SYS_TUNER_PATH).unwrap()
```
Comment (mvines) [inline sys-tuner/src/main.rs]: Why not just always attempt to `remove_file()` before binding

Comment (mvines) [reply sys-tuner/src/main.rs]: Also these unwraps() should be handled with a nice error eprintln!() and a clean `exit(1)`

File: sys-tuner/src/main.rs [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+        }
+    };
+
+    for stream in listener.incoming() {
+        if stream.is_ok() {
```
Comment (mvines) [inline sys-tuner/src/main.rs]: For monitoring it'd be nice to emit a log whenever a new connection occurs

File: sys-tuner/src/main.rs [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+        .arg("-eT")
+        .output()
+        .expect("Expected to see all threads");
+
+    if output.status.success() {
+        if let Ok(threads) = from_utf8(&output.stdout) {
+            let mut threads: Vec<&str> = threads.split('\n').collect();
+            for t in threads.iter_mut() {
+                if t.find("solana-poh-ser").is_some() {
+                    let pids: Vec<&str> = t.split_whitespace().collect();
```
Comment (mvines) [inline sys-tuner/src/main.rs]: A log here would be nice too that includes the pid/tid and whether the operation was successful or not

File: sys-tuner/src/main.rs [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+pub const SOLANA_SYS_TUNER_PATH: &str = "\0/tmp/solana-sys-tuner";
+
+#[allow(dead_code)]
+#[cfg(target_os = "linux")]
+fn tune_system() {
+    use std::process::Command;
+    use std::str::from_utf8;
+    use thread_priority::*;
+
+    let output = Command::new("ps")
```
Comment (mvines) [inline sys-tuner/src/main.rs]: It sorta sucks to rely on `ps` here.   How about just enumerate all directories in `/proc`, then read/parse `/proc/N/status`.  The first line in this file is "Name: solana-poh-serv`

File: sys-tuner/src/main.rs [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+    }
+}
+
+#[allow(dead_code)]
+#[cfg(any(not(unix), target_os = "macos"))]
+fn tune_system() {}
+
+#[allow(dead_code)]
+fn main() {
+    solana_logger::setup();
```
Comment (mvines) [inline sys-tuner/src/main.rs]: We should restrict who can use this daemon better.  For example right now I could get real time priority for my program, even if I'm not a `solana-validator` process by naming a thread "solana-poh-serv".  

There could be a command-line argument to sys-tuner that indicates the expected uid of the `solana-validator` process and ignore all "solana-poh-serv" threads that don't belong to that user, to at least slightly lock this down

File: net/remote/remote-node.sh (hunk 1 above)
Comment (pgarg66) [reply net/remote/remote-node.sh]: Done

File: sys-tuner/src/lib.rs (hunk 2 above)
Comment (pgarg66) [reply sys-tuner/src/lib.rs]: Won't this also trigger the rmem/wmem related tweaks? It does make more sense to call it `request_realtime_poh` from `poh_service` perspective.

File: sys-tuner/src/main.rs (hunk 5 above)
Comment (pgarg66) [reply sys-tuner/src/main.rs]: The compiler was complaining about it, since it's getting built for lib as well.

File: sys-tuner/src/main.rs (hunk 6 above)
Comment (pgarg66) [reply sys-tuner/src/main.rs]: `remove_file` does not work for abstract namespace sockets. This failure only happens when we run it on non-linux targets.

File: sys-tuner/src/main.rs (hunk 8 above)
Comment (pgarg66) [reply sys-tuner/src/main.rs]: done

File: sys-tuner/src/main.rs (hunk 7 above)
Comment (pgarg66) [reply sys-tuner/src/main.rs]: done

File: sys-tuner/src/main.rs (hunk 9 above)
Comment (pgarg66) [reply sys-tuner/src/main.rs]: I'll take a look at it.

File: sys-tuner/src/main.rs (hunk 10 above)
Comment (pgarg66) [reply sys-tuner/src/main.rs]: Abstract namespace sockets don't manifest in the filesystem. So, we cannot set file permissions for it. Actually, I had implemented the code to tweak the file permissions, but it failed on Linux for above reason. I can use regular (non-Abstract) sockets. That'll allow us to set correct group permissions.

Comment (mvines) [reply sys-tuner/src/main.rs]: Oh!  How does one set permission on an abstract namespace socket?

Comment (pgarg66) [reply sys-tuner/src/main.rs]: One way is to check the peers UID/GID after they connect to the socket. But, that's post connection.

Comment (mvines) [reply sys-tuner/src/main.rs]: That would be fine

Comment (pgarg66) [reply sys-tuner/src/main.rs]: For now, I have switched to using named socket. That also cleaned up the remove old socket workflow.

**Pull request merged**